### PR TITLE
Introduce FEEvaluationImplSelector

### PR DESCRIPTION
**Pull request by peterrum**
to remove the duplicated logic

### CODE REVIEW

File: include/deal.II/matrix_free/evaluation_selector.h [hunk 1]
```diff
@@ -76,7 +76,7 @@ SelectEvaluator<dim, fe_degree, n_q_points_1d, Number>::evaluate(
 {
   Assert(fe_degree >= 0 && n_q_points_1d > 0, ExcInternalError());
 
-  internal::FEEvaluationImplEvaluateSelector<dim, Number>::template run<
+  internal::FEEvaluationImplSelector<dim, Number, false>::template run<
```
[the remark this answers is not in the document]
Comment (kronbichler) [reply include/deal.II/matrix_free/evaluation_selector.h:79]: This was introduced in #4701 - it was done earlier than the refactoring into the precompiled `Impl` strategies, so these two concepts coexisted for a while, before #10922 and #13056 reorganized the code and made this class less important.

File: include/deal.II/matrix_free/evaluation_kernels.h [hunk 2]
```diff
@@ -2261,146 +2261,16 @@ namespace internal
    * currently supports $0\leq fe\_degree \leq 9$ and $degree+1\leq
    * n\_q\_points\_1d\leq fe\_degree+2$.
    */
-  template <int dim, typename Number>
-  struct FEEvaluationImplEvaluateSelector
+  template <int dim, typename Number, bool do_integrate>
+  struct FEEvaluationImplSelector
   {
-    template <int fe_degree, int n_q_points_1d>
+    template <int fe_degree, int n_q_points_1d, typename VNumber>
```
Comment (kronbichler) [inline include/deal.II/matrix_free/evaluation_kernels.h:2267]: What is `VNumber` and why is it needed? This should be documented; if needed, I suggest to use a more common naming scheme, such as `Number2` or `OtherNumber`.

**Review by kronbichler — APPROVED** [review]
This looks OK to, even though the saving of around 60 lines of code (plus some comments) is gained by another indirection.

File: include/deal.II/matrix_free/evaluation_kernels.h [hunk 3]
```diff
@@ -2261,147 +2261,24 @@ namespace internal
...
         FEEvaluationData<dim, Number, false> & fe_eval,
-        const bool                             sum_into_values_array)
+        const bool                             sum_into_values_array = false)
     {
+      // `OtherNumber` is either `const Number` (evaluate()) or `Number`
+      // (integrate())
+      static_assert(
+        std::is_same<Number,
+                     typename std::remove_const<OtherNumber>::type>::value,
+        "Type of Number and of OtherNumber do not match.");
```
Comment (peterrum) [inline include/deal.II/matrix_free/evaluation_kernels.h:2280]: @kronbichler  I have added a comment/assert that `OtherNumber` is either a const or non-const version of `Number`.

**Review by kronbichler — APPROVED** [review]

**Pull request merged**